Changed default to CleanSession(false) to preserve previous subscriptions

kamermans opened the pull request:
See issue #17 : upon reconnection, all existing subscriptions are lost.  The two ways I've found to fix this are to set `CleanSession(false)` in the MQTT client, or to explicitly resubscribe to all channels in the `OnConnect()` handler.  I don't see a good reason for the session to be cleaned during a reconnection, so I've added this PR to change the default of MQTT `CleanSession` from `true` to `false`.  I've also added an option to allow the user to change it back with `WithCleanSession()`.

Related discussion and docs:
https://github.com/eclipse/paho.mqtt.golang/issues/22
https://godoc.org/github.com/eclipse/paho.mqtt.golang#ClientOptions.SetCleanSession

Florimond commented:
Does it subscribes again using the exact same channel string, including options, like "last"?
I suppose it does, which can be good or bad depending on the use case...

kamermans commented:
Good question!  The paho documentation says that it "does not tell the server to clear the session when you reconnect", so the idea is that the MQTT server already knows your session state (based on your UUID in the case of emitter-go), and is willing to resume it.  Whether this means `last` is preserved, I don't know, some testing may be required to determine that.

Florimond commented:
Yes, that's what I wrote in the edited version of my answer.
Unless I missed something, Emitter doesn't implement anything regarding this flag on its side. To me, the way to go is to handle both subscription and re-subscription in the connection handler.

kamermans commented:
That is true, and I am on the fence here as well.  Personally, I've moved my `Subscribe()` functions into my `OnConnect()` handlers to be sure I'm subscribed every time I (re)connect.

The main problem I have with the current implementation is that it's basically impossible to guarantee that you are ever subscribed.  A microsecond after you subscribe, there could be a network interruption that triggers an auto-reconnection, then you're connected but not subscribed anymore.

Currently I'm using emitter across about 40 servers in three continents, and they never fail to find a way to disconnect, even though I've spread emitter cluster members around the earth to reduce latency.

Florimond commented:
I don't see it... If you set the OnConnect handler before sending your first connection request, and only subscribe in this handler: how then could you "miss" the subscription?

kamermans commented:
Sorry, I was referring to my code before I used OnConnect.  Now it is working properly.

MichaelHipp commented:
Shouldn't some of this be in the documentation, so every person who comes along doesn't have to re-discover it for themselves?

kamermans commented:
Definitely, although if this is just a documentation issue, we'll need to ensure that basically every code example uses this paradigm.

kamermans commented:
Also, it's probably still nice to be able to set `CleanSession()`, so if you don't want to turn it on by default, I can just change the default to `true` in this PR and people can set it to `false` if they so choose.

MichaelHipp commented:
IMHO, it should be set to false by default. That seems like the "least surprise" option to me. But in any event it needs to be documented.

Florimond commented:
I haven't tested the clean session set to false in Go, but I just did a quick test with the Python version and it doesn't re-subscribes. Which is in line with what I read [here about the Python lib](http://www.steves-internet-guide.com/mqtt-clean-sessions-example/) .

> the subscription information is remembered by the broker, and so the client doesn’t need to subscribe again

And about MQTT in general, on the [IBM knowledge center](https://www.ibm.com/support/knowledgecenter/en/SSFKSJ_7.1.0/com.ibm.mq.doc/tt60370_.htm) :

> If you set MqttConnectOptions.cleanSession to false before connecting, any subscriptions the client creates are added to all the subscriptions that existed for the client before it connected. All the subscriptions remain active when the client disconnects.

To me, it's clear everything is supposed to happen on the broker side. And the fact that the Go library takes upon itself to re-subscribe is a exceptional behavior. It does not follow the rule.

I think that the best we could do is to add to every emitter lib some "resubscribe" parameter, and implement the behavior unrelated to the MQTT's clean session flag.

kamermans commented:
Thanks for the response!

> the fact that the Go library takes upon itself to re-subscribe is a exceptional behavior. It does not follow the rule.

Agreed that we probably don't want to resubscribe explicitly in the emitter wrapper.  The argument for `CleanSession=true` would be that the emitter client already makes some non-default choices, like always passing a randomly generated user ID.

I'm understanding you correctly, you agree that the default behavior is that the user must explicitly re-subscribe to their channel(s) in `Connect()` or set `CleanSession()` to `false` (via this PR) to ensure a persistent subscription; and that the default behavior should be as it is today to avoid the Go client from diverging from the de-facto standard (and the other official Emitter clients).  That seems reasonable to me, despite the weirdness.  One argument for your position is that you really don't want a ton of persistent connections to be tracked by the broker from temporary clients.

It seems like this PR is still valuable, but I should change the default back to `true`, and add some documentation to make it clear that long-lived / persistent clients should set `CleanSession(false)` or resubscribe in `OnConnect()`.

Florimond commented:
> It seems like this PR is still valuable, but I should change the default back to true, and add some documentation to make it clear that long-lived / persistent clients should set CleanSession(false) or resubscribe in OnConnect().

Ok. Let's do this. I'll merge the pull request then...

kamermans commented:
Ok, sounds good.  I'll get it changed and let you know.

kamermans commented:
@Florimond @MichaelHipp I've updated the PR to maintain the current options and left the ability for users to set `cleanSession=false` if they so choose.  I've also added some prominent notes and examples to the README that should draw attention to the fact that sessions are not automatically resubscribed after a reconnection.

What do you think?

kelindar commented on README.md:
````diff
@@ -58,6 +61,39 @@ func main() {
+})
+```
+
+#### Method 2: Use Persistent Connections:
````
I don't think we should be mentioning this here, given that Emitter doesn't support it (and this is an emitter-specific library).

MichaelHipp replied:
Could you elaborate? Are you saying the documentation shouldn't give users a hint on how to achieve persistent connections? I think the docs ought to be as helpful as reasonably possible. Apologies if I misunderstood you.

kelindar replied:
The `cleanSession` flag work on traditional MQTT brokers, but it's not implemented in Emitter broker. So if you mention persistent connections in here, people would think that it actually works with Emitter.

kamermans replied:
Oh, thanks for the clarification @kelindar, I didn't realize that it doesn't work in Emitter.  Perhaps there is no point in adding `WithCleanSession()` either then?  I definitely think the first method (subscribing in `OnConnect()`) is important, otherwise there is no way to be sure you are even subscribed a second later since it silently reconnects without resubscribing.  This problem is at the root of the PR since I have found that running dozens of clients throughout the world, with an AWS ALB in front of emitter caused clients to disconnect occasionally, until eventually there were no clients left at all.

Florimond replied:
Well, that's what we discussed from the start. It seems you wanted to add this `WithCleanSession` anyway, because the go library actually resubscribe - according to the sources you cited. But, as I said earlier, it's specific to the go library, and the usual pattern should be to subscribe in the `OnConnect`

kamermans commented:
My understanding is that all MQTT clients resubscribe like this since it's a core feature of MQTT.  I did not think it was go-specific.
